File: plugin/src/main/java/com/coach/plugin/encounter/SchemaValidator.java
package com.coach.plugin.encounter;

import com.coach.plugin.encounter.model.BossDefinition;
import com.coach.plugin.encounter.model.CalloutDefinition;
import com.coach.plugin.encounter.model.EncounterPack;
import com.coach.plugin.encounter.model.MechanicDefinition;
import com.coach.plugin.encounter.model.PhaseDefinition;
import com.coach.plugin.encounter.model.TriggerDefinition;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Validates a parsed encounter pack against the schema v1 rules
 * (master-architecture §10). Returns all violations found — empty list = valid.
 *
 * Full JSON-Schema checking lives in the AI pipeline; this is the plugin's
 * load-time gate.
 */
public final class SchemaValidator
{
	private static final Set<String> TRIGGER_TYPES = Set.of(
		"animation", "projectile", "graphic", "npc_spawn", "npc_despawn",
		"hp", "tick_timer", "player_state", "location", "shout", "wave_cleared",
		"custom", "composite");

	private static final Set<String> CALLOUT_CATEGORIES = Set.of(
		"critical", "warning", "info", "transition");

	private static final Set<String> CONDITION_TYPES = Set.of(
		"npc_hp_below", "npc_hp_above", "player_hp_below", "player_hp_above",
		"tick_mod", "player_in_region", "prayer_active", "prayer_inactive",
		"inventory_contains", "custom");

	private static final int MIN_TICK_OFFSET = -5;
	private static final int MAX_TICK_OFFSET = 10;

	private List<String> errors = new ArrayList<>();

	/**
	 * Validate a pack. Stateless across calls: each invocation starts fresh.
	 */
	public List<String> validate(EncounterPack pack)
	{
		errors = new ArrayList<>();
		if (pack.schemaVersion == null)
		{
			errors.add("missing required field: schemaVersion");
			return errors;
		}
		if (!EncounterPack.SUPPORTED_SCHEMA_VERSION.equals(pack.schemaVersion))
		{
			errors.add("unsupported schemaVersion: '" + pack.schemaVersion
				+ "' (supported: " + EncounterPack.SUPPORTED_SCHEMA_VERSION + ")");
			return errors;
		}

		if (pack.metadata == null)
		{
			errors.add("missing required section: metadata");
		}
		else
		{
			requireText(pack.metadata.packId, "metadata.packId");
			requireText(pack.metadata.name, "metadata.name");
			requireText(pack.metadata.version, "metadata.version");
			requireText(pack.metadata.gameVersion, "metadata.gameVersion");
			if (pack.metadata.dependencies != null)
			{
				int i = 0;
				for (String dependency : pack.metadata.dependencies)
				{
					if (dependency == null || dependency.trim().isEmpty())
					{
						errors.add("metadata.dependencies[" + i + "] must be a non-empty packId");
					}
					i++;
				}
			}
		}

		if (pack.bosses == null || pack.bosses.isEmpty())
		{
			errors.add("pack must contain at least one boss");
			return errors;
		}

		Set<String> bossIds = new HashSet<>();
		for (BossDefinition boss : pack.bosses)
		{
			validateBoss(boss, bossIds);
		}
		return errors;
	}

	private void validateBoss(BossDefinition boss, Set<String> bossIds)
	{
		String label = bossLabel(boss);

		if (!requireText(boss.bossId, label + ".bossId")) return;
		if (!bossIds.add(boss.bossId))
		{
			errors.add("duplicate bossId: " + boss.bossId);
		}
		label = "boss[" + boss.bossId + "]";

		requireText(boss.name, label + ".name");
		requireNonNull(boss.npcId, label + ".npcId");

		if (boss.phases == null || boss.phases.isEmpty())
		{
			errors.add(label + ": at least one phase is required");
			return;
		}

		Set<String> phaseIds = new HashSet<>();
		for (PhaseDefinition phase : boss.phases)
		{
			validatePhase(boss.bossId, phase, phaseIds);
		}

		// shared boss-level mechanics have their own id scope
		validateMechanics(boss.bossId, boss.mechanics, label + ".mechanics");
	}

	private void validatePhase(String bossId, PhaseDefinition phase, Set<String> phaseIds)
	{
		String label = "boss[" + bossId + "] phase[" + (phase.phaseId != null ? phase.phaseId : "?") + "]";
		phase.bossId = bossId;

		if (!requireText(phase.phaseId, label + ".phaseId")) return;
		if (!phaseIds.add(phase.phaseId))
		{
			errors.add("duplicate phaseId '" + phase.phaseId + "' in boss " + bossId);
		}
		requireText(phase.name, label + ".name");

		if (phase.entryTrigger == null)
		{
			errors.add(label + ": entryTrigger is required");
		}
		else
		{
			validateTrigger(phase.entryTrigger, label + ".entryTrigger");
		}

		if (phase.exitTriggers != null)
		{
			int i = 0;
			for (TriggerDefinition trigger : phase.exitTriggers)
			{
				validateTrigger(trigger, label + ".exitTriggers[" + i++ + "]");
			}
		}

		validateMechanics(bossId, phase.mechanics, label + ".mechanics");
	}

	private void validateMechanics(String bossId, List<MechanicDefinition> mechanics, String label)
	{
		if (mechanics == null)
		{
			return;
		}
		// id uniqueness is scoped per mechanic list (one phase or the shared
		// boss-level list) — generated wave packs legitimately reuse attack
		// mechanics across phases, and runtime only evaluates the active phase.
		Set<String> scopeIds = new HashSet<>();
		for (MechanicDefinition mechanic : mechanics)
		{
			mechanic.bossId = bossId;
			String mLabel = label + "[" + (mechanic.mechanicId != null ? mechanic.mechanicId : "?") + "]";

			if (!requireText(mechanic.mechanicId, mLabel + ".mechanicId")) continue;
			if (!scopeIds.add(mechanic.mechanicId))
			{
				errors.add("duplicate mechanicId '" + mechanic.mechanicId + "' in boss " + bossId);
			}
			mLabel = label + "[" + mechanic.mechanicId + "]";
			requireText(mechanic.name, mLabel + ".name");

			if (mechanic.triggers == null || mechanic.triggers.isEmpty())
			{
				errors.add(mLabel + ": at least one trigger is required");
			}
			else
			{
				int i = 0;
				for (TriggerDefinition trigger : mechanic.triggers)
				{
					trigger.contextId = mechanic.mechanicId;
					validateTrigger(trigger, mLabel + ".triggers[" + i++ + "]");
				}
			}

			if (mechanic.callouts != null)
			{
				Set<String> calloutIds = new HashSet<>();
				int i = 0;
				for (CalloutDefinition callout : mechanic.callouts)
				{
					callout.bossId = bossId;
					validateCallout(callout, mLabel + ".callouts[" + i++ + "]", calloutIds);
				}
			}

			if (mechanic.cooldown != null && mechanic.cooldown < 0)
			{
				errors.add(mLabel + ".cooldown must be >= 0");
			}

			if (mechanic.conditions != null)
			{
				int i = 0;
				for (com.coach.plugin.encounter.model.ConditionDefinition condition : mechanic.conditions)
				{
					validateCondition(condition, mLabel + ".conditions[" + i++ + "]");
				}
			}
		}
	}

	private void validateCondition(com.coach.plugin.encounter.model.ConditionDefinition condition, String label)
	{
		if (!requireText(condition.type, label + ".type")) return;
		if (!CONDITION_TYPES.contains(condition.type))
		{
			errors.add(label + ": unknown condition type '" + condition.type + "'");
			return;
		}
		if ((condition.type.endsWith("_hp_below") || condition.type.endsWith("_hp_above"))
			&& condition.threshold == null)
		{
			errors.add(label + ": hp conditions require a threshold");
		}
		if ("tick_mod".equals(condition.type) && (condition.mod == null || condition.mod < 1))
		{
			errors.add(label + ": tick_mod requires mod >= 1");
		}
	}

	private void validateCallout(CalloutDefinition callout, String label, Set<String> calloutIds)
	{
		if (!requireText(callout.calloutId, label + ".calloutId")) return;
		if (!calloutIds.add(callout.calloutId))
		{
			errors.add("duplicate calloutId '" + callout.calloutId + "'");
			return;
		}
		label = label.replace("[?]", "[" + callout.calloutId + "]");
		requireText(callout.text, label + ".text");

		if (!requireText(callout.category, label + ".category")) return;
		if (!CALLOUT_CATEGORIES.contains(callout.category))
		{
			errors.add(label + ".category must be one of " + CALLOUT_CATEGORIES
				+ " but was '" + callout.category + "'");
		}

		if (callout.priority != null && (callout.priority < 1 || callout.priority > 100))
		{
			errors.add(label + ".priority must be 1-100");
		}

		checkTickOffset(callout.audioOffset, label + ".audioOffset");
		checkTickOffset(callout.visualOffset, label + ".visualOffset");

		if (callout.visual != null && callout.visual.type != null)
		{
			// visual type whitelist is enforced when rendering lands (Sprint 9+)
		}
	}

	private void validateTrigger(TriggerDefinition trigger, String label)
	{
		if (!requireText(trigger.type, label + ".type")) return;
		if (!TRIGGER_TYPES.contains(trigger.type))
		{
			errors.add(label + ": unknown trigger type '" + trigger.type + "'");
		}
		if ("composite".equals(trigger.type))
		{
			if (trigger.children == null || trigger.children.isEmpty())
			{
				errors.add(label + ": composite trigger requires children");
			}
			else if (!"AND".equalsIgnoreCase(trigger.logic) && !"OR".equalsIgnoreCase(trigger.logic))
			{
				errors.add(label + ": composite logic must be AND or OR");
			}
			else
			{
				int i = 0;
				for (TriggerDefinition child : trigger.children)
				{
					validateTrigger(child, label + ".children[" + i++ + "]");
				}
			}
		}
	}

	private void checkTickOffset(Integer offset, String label)
	{
		if (offset != null && (offset < MIN_TICK_OFFSET || offset > MAX_TICK_OFFSET))
		{
			errors.add(label + " must be between " + MIN_TICK_OFFSET + " and "
				+ MAX_TICK_OFFSET + " but was " + offset);
		}
	}

	private boolean requireText(String value, String field)
	{
		if (value == null || value.trim().isEmpty())
		{
			errors.add("missing required field: " + field);
			return false;
		}
		return true;
	}

	private boolean requireNonNull(Object value, String field)
	{
		if (value == null)
		{
			errors.add("missing required field: " + field);
			return false;
		}
		return true;
	}

	private static String bossLabel(BossDefinition boss)
	{
		return "boss[" + (boss.bossId != null ? boss.bossId : "?") + "]";
	}
}

File: knowledge-pipeline/src/schema_validator.py
"""Schema v1.0 structural validation — Python mirror of the plugin's
SchemaValidator rules (master-architecture §10), so drafts reach human review
already loadable by the plugin.

Every violation becomes a critical Issue with a stable code the AI validator
agent can auto-fix where safe.
"""
from __future__ import annotations

from typing import Any, Dict, List, Set

from validation_report import Issue, Report, SEV_CRITICAL

SCHEMA_VERSION = "1.0"
KNOWN_TRIGGER_TYPES = {
    "animation", "projectile", "graphic", "npc_spawn", "npc_despawn",
    "hp", "tick_timer", "player_state", "location", "shout", "wave_cleared",
    "custom", "composite",
}
CATEGORIES = {"critical", "warning", "info", "transition"}
MIN_OFFSET, MAX_OFFSET = -5, 10


def validate(draft: Dict[str, Any]) -> Report:
    report = Report()
    if draft.get("schemaVersion") != SCHEMA_VERSION:
        report.add(Issue(SEV_CRITICAL, "SCHEMA_VERSION",
                         f"schemaVersion must be '{SCHEMA_VERSION}'"))
        return report

    metadata = draft.get("metadata") or {}
    for field in ("packId", "name", "version", "gameVersion"):
        if not metadata.get(field):
            report.add(Issue(SEV_CRITICAL, "MISSING_FIELD",
                             f"metadata.{field} is required", path=f"metadata.{field}"))

    bosses = draft.get("bosses") or []
    if not bosses:
        report.add(Issue(SEV_CRITICAL, "NO_BOSSES", "at least one boss required"))
        return report

    seen_pack_bosses: Set[str] = set()
    for boss in bosses:
        _validate_boss(boss, seen_pack_bosses, report)
    return report


def _validate_boss(boss: Dict[str, Any], seen_pack_bosses: Set[str], report: Report) -> None:
    bid = boss.get("bossId")
    base = f"bosses[{bid if bid else '?'}]"
    if not bid:
        report.add(Issue(SEV_CRITICAL, "MISSING_FIELD", "bossId required", base))
        return
    if bid in seen_pack_bosses:
        report.add(Issue(SEV_CRITICAL, "DUPLICATE_ID",
                         f"duplicate bossId '{bid}'", base))
    seen_pack_bosses.add(bid)
    if not boss.get("name"):
        report.add(Issue(SEV_CRITICAL, "MISSING_FIELD", "boss name required", base))

    phases = boss.get("phases") or []
    if not phases:
        report.add(Issue(SEV_CRITICAL, "NO_PHASES", "at least one phase required",
                         base))
        return

    seen_phase_ids: Set[str] = set()
    for index, phase in enumerate(phases):
        path = f"{base}.phases[{index}]"
        pid = phase.get("phaseId")
        if not pid:
            report.add(Issue(SEV_CRITICAL, "MISSING_FIELD", "phaseId required", path))
            continue
        if pid in seen_phase_ids:
            report.add(Issue(SEV_CRITICAL, "DUPLICATE_ID",
                             f"duplicate phaseId '{pid}'", path))
        seen_phase_ids.add(pid)
        if not phase.get("entryTrigger"):
            report.add(Issue(SEV_CRITICAL, "MISSING_ENTRY_TRIGGER",
                             "phase entryTrigger required", path))

        seen_mechanics: Set[str] = set()
        seen_callout_ids_in_phase: Set[str] = set()
        for m_index, mechanic in enumerate(phase.get("mechanics") or []):
            _validate_mechanic(mechanic, f"{path}.mechanics[{m_index}]",
                               seen_mechanics, seen_callout_ids_in_phase, report)

    # shared boss-level mechanics have their own id scope
    seen_shared: Set[str] = set()
    seen_callouts_shared: Set[str] = set()
    for m_index, mechanic in enumerate(boss.get("mechanics") or []):
        _validate_mechanic(mechanic, f"{base}.shared.mechanics[{m_index}]",
                           seen_shared, seen_callouts_shared, report)


def _validate_mechanic(mechanic: Dict[str, Any], path: str,
                       seen_mechanics: Set[str],
                       seen_callout_ids: Set[str],
                       report: Report) -> None:
    mid = mechanic.get("mechanicId")
    mech_path = f"{path}[{mid if mid else '?'}]"
    if not mid:
        report.add(Issue(SEV_CRITICAL, "MISSING_FIELD",
                         "mechanicId required", mech_path))
        return
    if mid in seen_mechanics:
        report.add(Issue(SEV_CRITICAL, "DUPLICATE_ID",
                         f"duplicate mechanicId '{mid}'", mech_path))
    seen_mechanics.add(mid)
    if not mechanic.get("triggers"):
        report.add(Issue(SEV_CRITICAL, "NO_TRIGGERS",
                         "at least one trigger required", mech_path))

    cooldown = mechanic.get("cooldown")
    if cooldown is not None and cooldown < 0:
        report.add(Issue(SEV_CRITICAL, "COOLDOWN_NEGATIVE",
                         f"cooldown {cooldown} must be >= 0", mech_path))

    for t_index, trigger in enumerate(mechanic.get("triggers") or []):
        _validate_trigger(trigger, f"{mech_path}.triggers[{t_index}]", report)

    callouts = mechanic.get("callouts") or []
    if not callouts:
        report.add(Issue(SEV_WARNING, "NO_CALLOUTS",
                         "mechanic has no callouts (informational only)", mech_path))

    for c_index, callout in enumerate(callouts):
        _validate_callout(callout, f"{mech_path}.callouts[{c_index}]",
                          seen_callout_ids, report)


def _validate_trigger(trigger: Dict[str, Any], path: str, report: Report) -> None:
    ttype = trigger.get("type")
    if ttype not in KNOWN_TRIGGER_TYPES:
        report.add(Issue(SEV_CRITICAL, "UNKNOWN_TRIGGER_TYPE",
                         f"unknown trigger type '{ttype}'", path))


def _validate_callout(callout: Dict[str, Any], path: str,
                      seen_callout_ids: Set[str], report: Report) -> None:
    cid = callout.get("calloutId")
    if not cid:
        report.add(Issue(SEV_CRITICAL, "MISSING_FIELD",
                         "calloutId required", path))
        return
    if cid in seen_callout_ids:
        report.add(Issue(SEV_CRITICAL, "DUPLICATE_ID",
                         f"duplicate calloutId '{cid}'", path))
    seen_callout_ids.add(cid)
    if not callout.get("text"):
        report.add(Issue(SEV_CRITICAL, "MISSING_TEXT",
                         "callout text required (also TTS source)", path))

    category = callout.get("category")
    if category not in CATEGORIES:
        report.add(Issue(SEV_CRITICAL, "BAD_CATEGORY",
                         f"category must be one of {sorted(CATEGORIES)}, "
                         f"got '{category}'", path))

    priority = callout.get("priority")
    if priority is not None and not (1 <= priority <= 100):
        report.add(Issue(SEV_CRITICAL, "PRIORITY_RANGE",
                         f"priority {priority} must be 1-100", path))

    for offset_field in ("audioOffset", "visualOffset"):
        offset = callout.get(offset_field)
        if offset is not None and not (MIN_OFFSET <= offset <= MAX_OFFSET):
            report.add(Issue(SEV_CRITICAL, "OFFSET_RANGE",
                             f"{offset_field} {offset} must be "
                             f"{MIN_OFFSET}..{MAX_OFFSET}", path))
